### PR TITLE
support `null`/`undefined` procedure results

### LINKED ISSUE #527: Mutations with no return value stall forever
When a `resolve` callback’s return type is either:

- `Promise<void>`
- `Promise<null>`
- `Promise<undefined`

tRPC doesn’t ever close the response stream, thus, stalling the client indefinitely.

The following code **doesn’t work:**

```ts
trpc.router<Context>()
  .mutation("update", {
    input: z.object({ id: z.string() }),
    resolve: async ({ ctx, input }): Promise<void> => {
      await updateSection(input);
      // `return undefined` or `return null` wouldn’t solve the case
    },
  });
```

But the following return value change **does the trick:**

```ts
trpc.router<Context>()
  .mutation("update", {
    input: z.object({ id: z.string() }),
    resolve: async ({ ctx, input }): Promise<{ [key: string]: never }> => {
      await updateSection(input);
      return {};
    },
  });
```

### PR DESCRIPTION
**Pull request by KATT**
closes #527

### CODE REVIEW

File: packages/client/src/internals/executeChain.ts [hunk 1]
```diff
@@ -17,7 +17,12 @@ export function executeChain<
   op: Operation<TInput>;
 }) {
   type TError = TRPCClientError<TRouter>;
-  const $result = observableSubject<TOutput | null, TError>(null);
+  const $result = observableSubject<
+    { type: 'init' } | { type: 'data'; data: TOutput },
+    TError
+  >({
+    type: 'init',
+  });
```
Comment (KATT) [inline packages/client/src/internals/executeChain.ts:25]: This was the culprit - when we initialize as call we started with an `observableSubject` of `null` and we also used this to distinguish if a request was pending or had results.

**Review by kripod — APPROVED** [review]

File: packages/server/test/websockets.test.ts [hunk 2]
```diff
@@ -221,7 +221,7 @@ test('$subscription()', async () => {
   wsClient.close();
 });
 
-test('$subscription() - server randomly stop and restart (this test might be flaky, try re-running)', async () => {
+test.skip('$subscription() - server randomly stop and restart (this test might be flaky, try re-running)', async () => {
```
Comment (simonedelmann) [inline packages/server/test/websockets.test.ts:224]: What's the reason for this change?

File: packages/server/test/index.test.tsx [hunk 3]
```diff
@@ -533,3 +536,36 @@ describe('createCaller()', () => {
     sub.start();
   });
 });
+
+// regression https://github.com/trpc/trpc/issues/527
+test('void mutation response', async () => {
+  const { client, close, wssPort, router } = routerToServerAndClient(
+    trpc
+      .router()
+      .mutation('undefined', {
```
Comment (simonedelmann) [inline packages/server/test/index.test.tsx:545]: I wonder if we should have a test for `return undefined` as well? Should be the same, but I am not sure if `void` always equals `undefined`.

File: packages/client/src/internals/executeChain.ts (hunk 1 above)
Comment (simonedelmann) [reply packages/client/src/internals/executeChain.ts:25]: I get why using `null` is not possible and I think you have found a working solution. I wonder if it's possible to make it a little more intuitive. We cannot use `undefined` instead of `null`, right?

**Review by simonedelmann — COMMENTED** [review]
I left some comments, would be curious to hear what both of you think. 

Sorry for brevity and less code  snippets. Smartphone/Train...

File: packages/server/test/index.test.tsx (hunk 3 above)
Comment (KATT) [reply packages/server/test/index.test.tsx:545]: `void` === `undefined`
